core/src/main/scala/kafka/coordinator/GroupCoordinator.scala

```diff
@@ -647,12 +652,20 @@ class GroupCoordinator(val brokerId: Int,
     if (group.is(AwaitingSync))
       resetAndPropagateAssignmentError(group, Errors.REBALANCE_IN_PROGRESS)
 
-    group.transitionTo(PreparingRebalance)
-    info("Preparing to restabilize group %s with old generation %s".format(group.groupId, group.generationId))
-
-    val rebalanceTimeout = group.rebalanceTimeoutMs
-    val delayedRebalance = new DelayedJoin(this, group, rebalanceTimeout)
     val groupKey = GroupKey(group.groupId)
+
+    val (delay, timeout) = if (group.currentState != InitialRebalance) {
+      group.transitionTo(PreparingRebalance)
+      (group.rebalanceTimeoutMs, time.milliseconds() + group.rebalanceTimeoutMs)
+    } else {
+      joinPurgatory.cancelForKey(groupKey) match {
```

Comment:
I'm wondering if we can simply this logic a bit with the delayed join operation:

0. add another function in `purgatory` that does not "tryComplete" but directly `watchForOperation`.

1. add a flag indicating if a new member has joined. 
2. upon receiving a join group for unknown group, use `watchForOperation` directly with the rebalance timeout as `min(initial-delay, rebalance timeout)`, so that `tryCompleteJoin` will not be called immediately.
3. in `tryCompleteJoin` which is called upon another member has joined, the check if group has all rejoined members will always pass, we just need to set the new-member joined flag.
4. in `onCompleteJoin` (it can be execute from either the operation expiration path of operation), check if the flag is set; if yes re-watch the operation, otherwise complete the join.

By doing this we do not need to call cancel for key, and also do not need the group lock as above; also we can collapse the two states (initial rebalance and preparing rebalance) into one.